### PR TITLE
Add license checks for engine management APIs

### PR DESCRIPTION
**Pull request by ioanatia**
Enables Engines API under a platinum or trial license.

### CODE REVIEW

**Review by kderusso — APPROVED** [review]
LGTM. Are there any tests needed?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java [hunk 1]
```diff
@@ -57,6 +57,7 @@ public class XPackLicenseState {
         messages.put(XPackField.DEPRECATION, new String[] { "Deprecation APIs are disabled" });
         messages.put(XPackField.UPGRADE, new String[] { "Upgrade API is disabled" });
         messages.put(XPackField.SQL, new String[] { "SQL support is disabled" });
+        messages.put(XPackField.ENTEPRISE_SEARCH, new String[] { "Engines and behavioural analytics will be disabled" });
```
Comment (kderusso) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:60]: Typo: Should be `XPackField.ENTERPRISE_SEARCH`.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearchTransportAction.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.XPackField;
+
+public abstract class EnterpriseSearchTransportAction<Request extends ActionRequest, Response extends ActionResponse> extends
+    HandledTransportAction<Request, Response> {
+
+    protected final XPackLicenseState licenseState;
+
+    public static final LicensedFeature.Momentary LICENSED_ENGINE_FEATURE = LicensedFeature.momentary(
```
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearchTransportAction.java:30]: nit: LICENSED_ENT_SEARCH_FEATURES now that analytics is part of the check?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearchTransportAction.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        }
+    }
+
+    protected abstract void doExecute(Request request, ActionListener<Response> listener);
+
+    private static ElasticsearchSecurityException newComplianceException(XPackLicenseState licenseState) {
+        String licenseStatus = licenseState.statusDescription();
+
+        ElasticsearchSecurityException e = new ElasticsearchSecurityException(
+            "Current license is non-compliant for engines. Current license is {}. "
```
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearchTransportAction.java:62]: for engines and behavioural analytics?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java (hunk 1 above)
Comment (kderusso) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:60]: Also, this is nitpicky, but are we standardizing on behavioral or behavioural for analytics?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearchTransportAction.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            listener.onFailure(newComplianceException(licenseState));
+        }
+    }
+
+    protected abstract void doExecute(Request request, ActionListener<Response> listener);
+
+    private static ElasticsearchSecurityException newComplianceException(XPackLicenseState licenseState) {
+        String licenseStatus = licenseState.statusDescription();
+
+        ElasticsearchSecurityException e = new ElasticsearchSecurityException(
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearchTransportAction.java:61]: Does this message need to be updated for cloud/enterprise as well?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/TransportDeleteEngineAction.java [hunk 5]
```diff
@@ -9,29 +9,30 @@
...
     @Inject
     public TransportDeleteEngineAction(
         TransportService transportService,
         ActionFilters actionFilters,
-        EngineIndexService engineIndexService
+        EngineIndexService engineIndexService,
+        XPackLicenseState licenseState
     ) {
-        super(DeleteEngineAction.NAME, transportService, actionFilters, DeleteEngineAction.Request::new);
+        super(DeleteEngineAction.NAME, transportService, actionFilters, DeleteEngineAction.Request::new, licenseState);
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/TransportDeleteEngineAction.java:30]: If users downgrade their license and want to delete an Engine they're no longer using, will they have a documented way to do this?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearchTransportAction.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.license.License;
+import org.elasticsearch.license.LicensedFeature;
+import org.elasticsearch.license.XPackLicenseState;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.XPackField;
+
+public abstract class EnterpriseSearchTransportAction<Request extends ActionRequest, Response extends ActionResponse> extends
```
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearchTransportAction.java:25]: Analytics will extend specific master actions class so we shouldn't rely on `HandledTransportAction`. Can we expose a static function instead that takes the XPackLicenseState as argument and performs the checks?

**Review by jimczi — COMMENTED** [review]
It looks good, I left one comment regarding the abstract class. Considering how the analytics action are developed, I think we need to use a static function to check the license.

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/action/TransportDeleteEngineActionTests.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.nullValue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class TransportDeleteEngineActionTests extends ESTestCase {
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/action/TransportDeleteEngineActionTests.java:30]: Thank you for putting these tests in! Should we add tests for when licenses are supported as well?

**Review by kderusso — COMMENTED** [review]
Overall LGTM, a few questions and notes.

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/action/TransportDeleteEngineActionTests.java (hunk 7 above)
Comment (ioanatia) [reply x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/action/TransportDeleteEngineActionTests.java:30]: I think the yml tests already cover for this use case - but I might be wrong.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearchTransportAction.java (hunk 6 above)
Comment (ioanatia) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/EnterpriseSearchTransportAction.java]: Okay - I made this a base class for Engine transport actions and extracted the license checks in a new `LicenseUtils` class that hopefully can be reused by analytics.

**Review by kderusso — APPROVED** [review]

**Review by jimczi — APPROVED** [review]
LGTM, thanks @ioanatia

**Pull request merged**